### ISSUE TITLE
cudaMallocPitch error!!!

### ISSUE THREAD

**Issue by avcs2080** [Celebrandil/CudaSift#17]
cv::imread("image.png",0).convertTo(img1,CV_32FC1);
this works well.... no issue

however, in my application the image is ROI of another image so I can't simply read it from file. So,
img2 = image2(ROI)
then, cv::cvtColor(img2,img2,CV_BGR2GRAY);
then, img2.convertTo(img2,CV_32FC1);

This causes segmentation fault. I also checked if they belong to the same data type.
img1.type() = img2.type() = 5. I also printed the cv::Mat, and are similar in size and content

The error seems to be caused by 

safeCall(cudaMallocPitch((void **)&d_data, (size_t*)&pitch, (size_t)(sizeof(float)*width), (size_t)height)); 
(note there are two *'s inside (void) in the above statement)
under cudaImage.cu

I can't get my head around this. Any such error in the past....?

**Comment by rayhou0710** [comment]
I feel you use the wrong '-arch=sm_35' to compile. Check your GPU and modify the Makefile accordingly.

**Comment by avcs2080** [comment]
thanks for the reply.... the '-arch=sm_35' was set appropriately in the settings. I'm trying to implement your cuda kernels on a large dataset... so the kernels have to be called inside a loop. Ideally I'd avoid using cuda kernels inside a loop but I have no choice. I have solved the problem but the solution seems a little unusual. I had to strictly follow the following steps

here is how
 img_1_host = cv::Mat::zeros(IMG_1_HOST.rows, IMG_1_HOST.cols,CV_32FC1);  // STEP1
 SiftData siftDataR;                                                                                              // STEP2
 InitSiftData(siftDataR, 65535, true, true);                                                          // STEP3
 CudaImage IMG_1_DEV; 
// declaring cudaimage inside the loop solves the 'OUT OF MEMORY PROBLEM'
cv::cvtColor(IMG_1_HOST,img_1_host,CV_BGR2GRAY);                    // STEP4
img_1_host.convertTo(img_1_host,CV_32FC1);                                  // STEP5
.......

**Comment by rremani** [comment]
Hi @avcs2080 ,
I have got out of memory error when I try to MatchSiftData() of nearly 20000 features from each image. Does it related to memory allocation error.
Thanks

**Comment by avcs2080** [comment]
@rremani 
It is hard to say without looking at the code. did you initialize SiftData with 20000  values prior to using it i.e.
something like
     SiftData siftDataA;  
     InitSiftData(siftDataA, 20000, true, true);
In my case, I had nested for loop as discussed above. When I declared CudaImage outside the nested loop, surprisingly I got out of memory error. Declaring it inside the loop somehow got rid of the error. I'm assuming the scope of the variable within the loop might have prevented it. Also,  don't forget to free the memory at the bottom.

hope this helps
